crates/matrix-sdk/src/live_location_share.rs

```diff
@@ -14,11 +14,53 @@
 
 //! Types for live location sharing.
 
+use async_stream::stream;
 use ruma::{
-    events::{beacon_info::BeaconInfoEventContent, location::LocationContent},
-    MilliSecondsSinceUnixEpoch, OwnedUserId,
+    events::{
+        beacon::OriginalSyncBeaconEvent, beacon_info::BeaconInfoEventContent,
+        location::LocationContent,
+    },
+    MilliSecondsSinceUnixEpoch, OwnedUserId, RoomId,
 };
 
+use crate::{event_handler::ObservableEventHandler, Client, Room};
+use futures_util::Stream;
+
+/// An observable live location.
+#[derive(Debug)]
+pub struct ObservableLiveLocation {
+    observable_room_events: ObservableEventHandler<(OriginalSyncBeaconEvent, Room)>,
+}
+
+impl ObservableLiveLocation {
+    /// Create a new `ObservableLiveLocation` for a particular room.
+    pub(crate) fn new(client: &Client, room_id: &RoomId) -> Self {
+        Self { observable_room_events: client.observe_room_events(room_id) }
+    }
+
+    /// Get a stream of [`LiveLocationShare`].
+    pub fn subscribe(&self) -> impl Stream<Item = LiveLocationShare> {
+        let stream = self.observable_room_events.subscribe();
+
+        stream! {
```

Comment:
Cannot use `StreamExt::map` because `room.get_user_beacon_info` is used inside this map, and is an async method. Instead, we use the `stream!` macro, which does the job too.